File: day22/src/main/java/de/havox_design/aoc2016/day22/Day22.java
package de.havox_design.aoc2016.day22;

import de.havox_design.aoc2016.utils.input.DataReader;

import java.util.List;

public class Day22 {
    private final List<String> input;

    public Day22(String fileName) {
        input = readData(fileName);
    }

    public static long solvePart1(String fileName) {
        Day22 instance = new Day22(fileName);
        return instance.solvePart1();
    }

    public static long solvePart2(String fileName) {
        Day22 instance = new Day22(fileName);
        return instance.solvePart2();
    }

    public long solvePart1() {
        return 0L;
    }

    public long solvePart2() {
        return 0L;
    }

    private List<String> readData(String fileName) {
        return DataReader.readData(fileName, MainClass.class);
    }
}

File: day22/src/main/kotlin/de/havox_design/aoc2016/day22/Day22.kt
package de.havox_design.aoc2016.day22

class Day22(private var filename: String) {
    fun solvePart1(): Int {
        val nodes = getResourceAsText(filename)
            .drop(2)
            .mapNotNull { parse(it) }
        return nodes
            .flatMap { node ->
                nodes.mapNotNull { other ->
                    setOf(node, other).takeIf { node != other && node.used > 0 && node.used <= other.available }
                }
            }
            .toSet()
            .size
    }

    fun solvePart2(): Long =
        0L

    private fun parse(line: String): Node? =
        "/dev/grid/node-x(?<x>\\d+)-y(?<y>\\d+)\\s+(?<size>\\d+)T\\s+(?<used>\\d+)T.+"
            .toRegex()
            .matchEntire(line)?.destructured
            ?.let { (x, y, size, used) -> Node(x.toInt(), y.toInt(), size.toInt(), used.toInt()) }

    private fun getResourceAsText(path: String): List<String> =
        this.javaClass.classLoader.getResourceAsStream(path)!!.bufferedReader().readLines()
}